enterprise/internal/campaigns/webhooks_gitlab_test.go

```diff
@@ -0,0 +1,1008 @@
+
+				rec := httptest.NewRecorder()
+				h.ServeHTTP(rec, req)
+
+				resp := rec.Result()
+				if have, want := resp.StatusCode, http.StatusNotImplemented; have != want {
+					t.Errorf("unexpected status code: have %d; want %d", have, want)
+				}
+				assertBodyIncludes(t, resp.Body, "unknown object kind")
+			})
+
+			t.Run("error from handleEvent", func(t *testing.T) {
+				store, rstore, clock := gitLabTestSetup(t, db)
+				h := NewGitLabWebhook(store, rstore, clock.now)
+				es := createGitLabExternalService(t, ctx, rstore)
+				repo := createGitLabRepo(t, ctx, rstore, es)
+				changeset := createGitLabChangeset(t, ctx, store, repo)
+				body := createMergeRequestPayload(t, repo, changeset, "close")
+
+				// Remove the URL from the GitLab configuration.
+				cfg, err := es.Configuration()
+				if err != nil {
+					t.Fatal(err)
+				}
+				conn := cfg.(*schema.GitLabConnection)
+				conn.Url = ""
+				es.Config = marshalJSON(t, conn)
+				if err := rstore.UpsertExternalServices(ctx, es); err != nil {
+					t.Fatal(err)
+				}
+
+				u := extsvc.WebhookURL(extsvc.TypeGitLab, es.ID, "https://example.com/")
+				req, err := http.NewRequest("POST", u, bytes.NewBufferString(body))
+				if err != nil {
+					t.Fatal(err)
+				}
+				req.Header.Add(webhooks.TokenHeaderName, "secret")
+
+				rec := httptest.NewRecorder()
+				h.ServeHTTP(rec, req)
+
+				resp := rec.Result()
+				if have, want := resp.StatusCode, http.StatusInternalServerError; have != want {
+					t.Errorf("unexpected status code: have %d; want %d", have, want)
+				}
+				assertBodyIncludes(t, resp.Body, "could not determine service id")
+			})
+
+			// The valid tests below are pretty "happy path": specific unit
+			// tests for the utility methods on GitLabWebhook are below. We're
+			// mostly just testing the routing here, since these are ServeHTTP
+			// tests; however, these also act as integration tests. (Which,
+			// considering they're ultimately invoked from TestIntegration,
+			// seems fair.)
+
+			t.Run("valid merge request approval events", func(t *testing.T) {
+				for _, action := range []string{"approved", "unapproved"} {
+					t.Run(action, func(t *testing.T) {
+						store, rstore, clock := gitLabTestSetup(t, db)
+						h := NewGitLabWebhook(store, rstore, clock.now)
+						es := createGitLabExternalService(t, ctx, rstore)
+						repo := createGitLabRepo(t, ctx, rstore, es)
+						changeset := createGitLabChangeset(t, ctx, store, repo)
+						body := createMergeRequestPayload(t, repo, changeset, "approved")
+
+						u := extsvc.WebhookURL(extsvc.TypeGitLab, es.ID, "https://example.com/")
+						req, err := http.NewRequest("POST", u, bytes.NewBufferString(body))
+						if err != nil {
+							t.Fatal(err)
+						}
+						req.Header.Add(webhooks.TokenHeaderName, "secret")
+
+						repoupdater.MockEnqueueChangesetSync = func(ctx context.Context, ids []int64) error {
+							if diff := cmp.Diff(ids, []int64{changeset.ID}); diff != "" {
```

Comment:
should we also assert this function is called?

```suggestion
                                                  changesetEnqueued := false
						repoupdater.MockEnqueueChangesetSync = func(ctx context.Context, ids []int64) error {
				                         changesetEnqueued = true
							if diff := cmp.Diff(ids, []int64{changeset.ID}); diff != "" {
```

Reply:
Yes. Yes, we should.